module4/lessons/jest_supertest_and_graphql.md

```diff
@@ -0,0 +1,213 @@
+title: Testing Graphql with Jest and Supertest
+---
+
+## Learning Goals
+- Understand how to set-up Jest & Supertest in an Express app and test GraphQL endpoints.
+- Understand a bit more about Crates architecture.
+
+## Warm Up
+- What are the steps necessary to writing a good test?
+- What are some differences between GraphQL and REST?
+- Are there any parallels that can be drawn between an express/Node application and a Rails application?
+
+## Setting up a Test DB
+
+- In Rails, we have `RAILS_ENV` which allows us to execute code in specific environments, similarly, Node has `NODE_ENV`. When we are running our specs we need to make sure that we are running against a `TEST` database, and not our `development` or `production` database. So, let's get started:
+```
+$ cd <path_to_crate>
+$ git checkout -b <a_new_testing_branch>
+```
+If you `cd` into the `crate/code/api/src/config` directory, you will notice a few files that configure our express application. The one we are interested in at this point is `database.json`. If you open up that file, you should see:
+```js
+{
+  "development": {
+    "username": "postgres",
+      "password": null,
+      "database": "crate",
+      "host": "127.0.0.1",
+      "dialect": "postgres",
+      "seederStorage": "sequelize"
+  },
+    "production": {
+      "username": "postgres",
+      "password": null,
+      "database": "crate",
+      "host": "127.0.0.1",
+      "dialect": "postgres",
+      "seederStorage": "sequelize"
+    }
+}
+```
+This file is where we setup our `development` and `production` databases; however, since Crate has no tests, we don't have a block for our `test` ENV. This can be problematic if we have specs that modify or delete data, since we don't want to touch or modify our development database. So, let's add another block:
+```js
+{
+  "development": {
+    "username": "postgres",
+      "password": null,
+      "database": "crate",
+      "host": "127.0.0.1",
+      "dialect": "postgres",
+      "seederStorage": "sequelize"
+  },
+  "test": {
+    "username": "postgres",
+      "password": null,
+      "database": "crate_testing", <== Note: This needs to be a different name than our development or production database, otherwise psql will get confused.
+      "host": "127.0.0.1",
+      "dialect": "postgres",
+      "seederStorage": "sequelize"
+  },
+    "production": {
+      "username": "postgres",
+      "password": null,
+      "database": "crate",
+      "host": "127.0.0.1",
+      "dialect": "postgres",
+      "seederStorage": "sequelize"
+    }
+}
+```
+Now that we have a test configuration in our `database.json`, let's create the database:
+```bash
+$ psql
+psql (13.1)
+Type "help" for help.
+
+<YOUR_NAME_OR_POSTGRES_USER_NAME>=# CREATE DATABASE crate_testing;
+CREATE DATABASE
+\l; # <= Check to make sure the DB was created
+<A LIST OF DATABASES>
+\q # <= Quit psql
+```
+Great! Now we have a test db, so let's add some data to it:
+
+```bash
+$ NODE_ENV=test npm setup:db
```

Comment:
I would add a small note around what is happening here with `NODE_ENV=test` in the npm command. Not sure they have seen environment variables used here like this.

Reply:
I needed to use run with this command to get it to execute. `NODE_ENV=test npm run setup:db`

Reply:
Updated!